packages/basic-crawler/src/internals/basic-crawler.ts

```diff
@@ -952,20 +953,30 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
             throw error;
         }
 
-        const shouldRetryRequest = !request.noRetry && request.retryCount < this.maxRequestRetries && !(error instanceof NonRetryableError);
+        const isUserRequestedRetry = error instanceof RetryRequest;
+
+        const shouldRetryRequest = isUserRequestedRetry
+            || (!request.noRetry && request.retryCount < this.maxRequestRetries && !(error instanceof NonRetryableError));
```

Comment:
This has become quite complex. Can we split it into multiple named variables so it reads better pls?